### DISCUSSION TITLE
Prevent execution of animation after dynamically changing variants

### DISCUSSION THREAD

**Discussion by timfuhrmann** [motiondivision/motion#1555] [Q&A]
I have a page transition that varies respective to the window's width. It works fine, but whenever `isDesktop` changes and the variants are being switched, the animation is being triggered, in other word a page transition on resize. Is there any way to prevent that, to exclusively trigger the animation on route/key change?

```typescript
const [windowWidth, setWindowWidth] = useState<number | null>(null);

const isDesktop = !!windowWidth && windowWidth > 1024;

useLayoutEffect(() => {
    setWindowWidth(window.innerWidth);
}, []);

useEffect(() => {
    const width = window.innerWidth;

    const onResize = () => {
        if (width !== window.innerWidth) {
            setWindowWidth(window.innerWidth);
        }
    };

    window.addEventListener("resize", onResize);
    return () => window.removeEventListener("resize", onResize);
}, []);

<AnimatePresence initial={false}>
    <StageWrapper
        key={router.asPath}
        initial="initial"
        animate="animate"
        variants={isDesktop ? wrapperVariants : mobileWrapperVariants}>
        <Stage
            key={router.asPath}
            transformTemplate={transform}
            initial="initial"
            animate="animate"
            exit="exit"
            variants={isDesktop ? stageVariants : undefined}>
            <StageFrame>
                <StageInner>{children}</StageInner>
            </StageFrame>
        </Stage>
    </StageWrapper>
</AnimatePresence>
```

**Comment by harshagarwalnyu** [comment]
The problem: `isDesktop` changes → React re-renders → new `variants` object → Framer Motion sees new values and plays the transition. It can't distinguish "route changed (animate)" from "window resized (don't animate)".

**Fix 1 — Use `useReducedMotion` / `initial={false}` guard**

Add a ref to track whether the variant change is from resize vs. route:

```tsx
const isFirstRender = useRef(true);
const prevIsDesktop = useRef(isDesktop);

useEffect(() => {
  prevIsDesktop.current = isDesktop;
}, [isDesktop]);

const shouldAnimate = !isFirstRender.current || routeChanged;
```

**Fix 2 — Keep variants stable, change only on route (recommended)**

Don't put `isDesktop` variants inside the component that animates. Instead:
```tsx
// Compute once outside render, or memoize
const variants = useMemo(() => ({
  initial: isDesktop ? { x: '-100%' } : { y: '-100%' },
  animate: { x: 0, y: 0 },
  exit: isDesktop ? { x: '100%' } : { y: '100%' },
}), [isDesktop]);

// Use layoutId for the size change, key for route change
<AnimatePresence mode="wait">
  <motion.div
    key={router.pathname}  // only route changes trigger exit
    variants={variants}
    initial="initial"
    animate="animate"
    exit="exit"
  />
</AnimatePresence>
```

When `isDesktop` changes but `router.pathname` stays the same, `AnimatePresence` sees no unmount/mount — no exit animation fires. The `variants` update mid-animation just adjusts direction, not replays from `initial`.

**Fix 3 — CSS for layout, Framer for route only**

Use CSS media queries for the direction difference and Framer only for opacity/scale transitions that are direction-agnostic. This cleanly separates concerns.
